src/main/java/com/amazon/ion/impl/macro/TemplateBodyExpression.kt

```diff
@@ -0,0 +1,81 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.amazon.ion.impl.macro
+
+import com.amazon.ion.IonType
+import com.amazon.ion.SymbolToken
+import com.amazon.ion.Timestamp
+import java.math.BigDecimal
+import java.math.BigInteger
+
+/**
+ * Represents an expression in the body of a template.
+ *
+ * We cannot use [`IonValue`](com.amazon.ion.IonValue) for this because `IonValue` requires references to parent
+ * containers and to an IonSystem which makes it impractical for reading and writing macros definitions. Furthermore,
+ * there is information we need to capture that cannot be expressed in the IonValue model, such as macro invocations
+ * and variable references.
+ *
+ * A template body is compiled into a list of expressions, without nesting, for ease and efficiency of evaluating
+ * e-expressions. Because of this, the container types do not have other values nested in them; rather they contain a
+ * range that indicates which of the following expressions are part of that container.
+ */
+sealed interface TemplateBodyExpression {
+    // TODO: Special Forms (if_void, for, ...)?
+
+    /**
+     * A temporary placeholder that is used only while a macro is partially compiled.
+     */
+    object Placeholder : TemplateBodyExpression
+
+    // Scalars
+    data class NullValue(val annotations: List<SymbolToken> = emptyList(), val type: IonType) : TemplateBodyExpression
+    data class BoolValue(val annotations: List<SymbolToken> = emptyList(), val value: Boolean) : TemplateBodyExpression
+    data class IntValue(val annotations: List<SymbolToken> = emptyList(), val value: Long) : TemplateBodyExpression
+    data class BigIntValue(val annotations: List<SymbolToken> = emptyList(), val value: BigInteger) : TemplateBodyExpression
+    data class FloatValue(val annotations: List<SymbolToken> = emptyList(), val value: Double) : TemplateBodyExpression
+    data class DecimalValue(val annotations: List<SymbolToken> = emptyList(), val value: BigDecimal) : TemplateBodyExpression
```

Comment:
Will probably have to be `Decimal` to support -0...

Reply:
`Decimal` extends `BigDecimal`, so this can be constructed with a `-0` decimal, but I can't think of any problem with changing it to `Decimal` here.

Reply:
So it ends up making the tests (or other programmatic construction of a template body) more annoying if I change this to `Decimal`. I'm going to leave it for now, and we can revisit it later if it causes problems.